diff -Naur aio_pika-9.4.1-old/queue.py aio_pika-9.4.1-new/queue.py
--- aio_pika-9.4.1-old/queue.py	2024-04-21 15:41:42.834610806 +0300
+++ aio_pika-9.4.1-new/queue.py	2024-04-25 15:13:09.707761029 +0300
@@ -484,6 +484,11 @@
         self._queue = asyncio.Queue()
         self._consume_kwargs = kwargs
 
+        if kwargs.get("timeout"):
+            self.__anext_impl = self.__anext_with_timeout
+        else:
+            self.__anext_impl = self.__anext_without_timeout
+
         self._amqp_queue.close_callbacks.add(self.close)
 
     async def on_message(self, message: AbstractIncomingMessage) -> None:
@@ -511,6 +516,9 @@
         await self.close()
 
     async def __anext__(self) -> IncomingMessage:
+        return await self.__anext_impl()
+
+    async def __anext_with_timeout(self) -> IncomingMessage:
         if not hasattr(self, "_consumer_tag"):
             await self.consume()
         try:
@@ -530,5 +538,14 @@
             await asyncio.wait_for(self.close(), timeout=timeout)
             raise

+    async def __anext_without_timeout(self) -> IncomingMessage:
+        if not hasattr(self, "_consumer_tag"):
+            await self.consume()
+        try:
+            return await self._queue.get()
+        except asyncio.CancelledError:
+            await self.close()
+            raise
+

 __all__ = ("Queue", "QueueIterator", "ConsumerTag")
